File: leetcode/SkipList.rs
/**********************************************
  > File Name		: SkipList.rs
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Mon 04 Apr 2022 06:54:51 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

use std::rc::Rc;
use std::cell::RefCell;
use rand::prelude::*;

const MAX_LEVEL: usize = 32;

struct SkipListNode<T> {
    val: T,
    layers: RefCell<Vec<Option<Rc<Self>>>>
}

impl<T> PartialEq for SkipListNode<T>
where T: Eq {
    fn eq(&self, other: &Self) -> bool {
        self.val == other.val
    }
}

impl<T> Eq for SkipListNode<T> where T: Ord {}

impl<T> Ord for SkipListNode<T>
where T: Ord {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.val.cmp(&other.val)
    }
}

impl<T> PartialOrd for SkipListNode<T>
where T: Ord {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl<T> SkipListNode<T> {
    fn new(val: T, size: usize) -> Self {
        Self {
            val,
            layers: RefCell::new(vec![None; size])
        }
    }

    #[inline]
    fn next(&self, layer: usize) -> Option<Rc<Self>> {
        self.layers.borrow().get(layer).unwrap().clone()
    }
}

struct SkipList<T> {
    level: usize,
    head: Rc<SkipListNode<T>>,
    tail: Rc<SkipListNode<T>>
}

impl SkipList<i32> {
    fn new() -> Self {
        let tail = Rc::new(SkipListNode::<i32>::new(std::i32::MAX, 0));
        let head = SkipListNode::<i32>::new(std::i32::MAX, MAX_LEVEL);
        {
            let mut head_layers = head.layers.borrow_mut();

            for i in 0..MAX_LEVEL {
                (*head_layers)[i] = Some(tail.clone());
            }
        }
            
        Self {
            level: 0,
            head: Rc::new(head),
            tail
        }
    }
}

impl<T> SkipList<T> where T: std::fmt::Display {
    fn random_level() -> usize {
        let mut res: usize = 1;
        let mut deno: u32 = 2;
        let mut rng = rand::thread_rng();
        
        while res < MAX_LEVEL && rng.gen_ratio(1, deno) {
            res += 1;
            deno <<= 1;
        }
        res
    }

    fn find_at_layer(&self, mut cursor: Rc<SkipListNode<T>>, layer: usize, target: &T) 
        -> Rc<SkipListNode<T>> where T: Ord {

        if layer >= self.level {return cursor;}

        loop {
            cursor = match cursor.next(layer) {
                Some(next_node) if &next_node.val < target => next_node.clone(),
                _ => break
            };
        }
        cursor
    }

    pub fn add(&mut self, val: T) where T: Ord {
        let mut update: Vec::<Option<Rc<SkipListNode<T>>>> = vec![None; MAX_LEVEL];
        let mut cursor = self.head.clone();

        let mut index = self.level;
        while index > 0 {
            index -= 1;
            cursor = self.find_at_layer(cursor, index, &val);
            update[index] = Some(cursor.clone());
        }

        let lv = {
            let mut lv = Self::random_level();
            if lv > self.level {
                self.level += 1;
                lv = self.level;
                update[lv - 1] = Some(self.head.clone());
            }
            lv
        };

        let new_node = Rc::new(SkipListNode::<T>::new(val, lv));
        let mut new_node_layers = new_node.layers.borrow_mut();
        index = lv;

        while index > 0 {
            index -= 1;
            let p = update.get(index).unwrap().as_ref().unwrap().clone();
            (*new_node_layers)[index] = match p.layers.borrow().get(index).unwrap() {
                None => None,
                Some(v) => Some(v.clone())
            };
            (*p.layers.borrow_mut())[index] = Some(new_node.clone());
        }
    }

    pub fn erase(&mut self, target: T) -> bool where T: Ord {
        let mut update: Vec<Option<Rc<SkipListNode<T>>>> = vec![None; MAX_LEVEL + 1];
        let mut cursor = self.head.clone();
        let mut i = self.level;

        while i > 0 {
            i -= 1;
            cursor = self.find_at_layer(cursor, i, &target);
            update[i] = Some(cursor.clone());
        }

        //cursor = cursor.clone().layers.borrow().get(0).unwrap().as_ref().unwrap().clone();
        cursor = cursor.next(0).unwrap();
        if cursor.val != target {
            return false;
        }

        i = 0;
        while i < self.level {
            match update.get(i).unwrap().as_ref().unwrap().layers.borrow().get(i).unwrap() {
                Some(skiplist_node) if skiplist_node == &cursor => {},
                _ => break
            }
            println!("layers length = {}", cursor.layers.borrow().len());
            (*update[i].as_ref().unwrap().layers.borrow_mut())[i] = match cursor.layers.borrow().get(i) {
                Some(node) if node.is_some() => node.clone(),
                _ => None
            };
            i += 1;
        }

        while self.level > 0 && match self.head.layers.borrow().get(self.level-1) {
            Some(node) if node.is_some() => node.as_ref().unwrap() == &self.tail,
            _ => false
        } { //head layers are always be Some.
            self.level -= 1;
        }
        true
    }

    pub fn query(&self, target: T) -> bool where T: Ord {
        let mut height = self.level;
        let mut cursor = self.head.clone();
        
        while height > 0 {
            height -= 1;
            cursor = self.find_at_layer(cursor, height, &target);
        }
        //cursor = cursor.clone().layers.borrow().get(0).unwrap().as_ref().unwrap().clone();
        cursor = cursor.next(0).unwrap();
        cursor.val == target
    }
}

fn main() {
    for i in 0..6 {
        println!("{}", i);
    }
}

File: leetcode/SkipList.cpp
/**********************************************
  > File Name		: SkipList.cpp
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Fri 08 Apr 2022 02:55:45 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

#include <list>
#include <vector>
#include <cstdlib>
#include <stdio.h>
#include <climits>
using namespace std;

template <typename T>
class Node {
public:
    T val;
    typename list<Node<T>>::iterator next_level_node;

    Node(T val): val(val) {}
    Node(T val, typename list<Node<T>>::iterator next_level_node): val(val), next_level_node(next_level_node) {}
};

/**
 * Allows add duplicate elements
 */
template <typename T>
class SkipList {
private:
    vector<list<Node<T>>> levels;
    vector<typename list<Node<T>>::iterator> nodes;

    static int random_level() {
        int level = 0;
        while (rand() < (RAND_MAX >> 1))
            level++;
        return level;
    }

    static void print_ll(const list<Node<T>>& ll) {
        if (ll.empty()) {
            printf("empty\n");
            return;
        }
        for (auto iter = ll.cbegin(); iter != ll.cend(); iter++) {
            printf("%d->", iter->val);
        }
        printf("\n");
    }

public:
    bool search(T target) {
        int level_index = this->levels.size()-1;
        const list<Node<T>>* level = &this->levels[level_index];
        auto iter = level->begin();

        while (level_index >= 0) {
            while (iter != level->end() && target > iter->val) {
                iter++;
            }

            if (iter != level->end() && target == iter->val) {
                return true;
            }

            if (!level_index)
                return false;

            if (iter == level->begin()) {
                level = &this->levels[--level_index];
                iter = level->begin();
            } else {
                iter--;
                level = &this->levels[--level_index];
                iter = iter->next_level_node;
            }

        }
        return false;
    }

    void add(T val) {
        int level_num = random_level();
        for (int i = 0, size = this->levels.size(); i < level_num - size + 1; i++) {
            this->levels.push_back(list<Node<T>>());
        }

        list<Node<T>>* level = &this->levels[level_num];
        auto iter = level->begin();

        while (level_num >= 0) {
            while (iter != level->end() && val > iter->val) {
                iter++;
            }

            ///uncomment these, duplicate code not allowed.
            //if (iter != level->end() && val == iter->val) {
                //this->nodes.push_back(iter);
                //break;
            //}

            level->emplace(iter, Node<T>(val));
            this->nodes.push_back(--iter);

            if (!level_num)
                break;

            if (iter == level->begin()) {
                level = &this->levels[--level_num];
                iter = level->begin();
            } else {
                level = &this->levels[--level_num];
                iter = (--iter)->next_level_node;
            }
        }

        for (int i = 0; i < this->nodes.size()-1; i++) {
            this->nodes[i]->next_level_node = this->nodes[i+1];
        }
        this->nodes.clear();
    }

    bool erase(T target) {
        int level_index = this->levels.size()-1;
        list<Node<T>>* level = &this->levels[level_index];
        auto iter = level->begin();

        while (level_index >= 0) {
            while (iter != level->end() && target > iter->val)
                iter++;

            if (iter != level->end() && target == iter->val) {
                auto temp = iter->next_level_node;
                level->erase(iter);

                if (!level_index) {
                    return true;
                }
                else {
                    level = &this->levels[--level_index];
                    iter = temp;
                }
            } else {
                if (!level_index) {
                    return false;
                }
                level = &this->levels[--level_index];
                iter = level->begin();
            }
        }

        return false;
    }

    void show_sl() {
        int level_num = this->levels.size()-1;
        
        while (level_num >= 0) {
            printf("level %d: ", level_num);
            print_ll(this->levels[level_num--]);
        }
    }
};

struct SkipListNode {
	int val;
	vector<SkipListNode *> level;
	SkipListNode (int _val, int sz=32) : val(_val), level(sz, nullptr) {}
};

class Skiplist {
private:
    SkipListNode *head, *tail;
    int level, length;
public:
	static constexpr int MAXL = 32;
    static constexpr int P = 4;
    static constexpr int S = 0xFFFF;
    static constexpr int PS = S / 4;

    Skiplist() {
        level = length = 0;
        tail = new SkipListNode(INT_MAX, 0);
        head = new SkipListNode(INT_MAX);
        for (int i = 0; i < MAXL; ++i) { 
        	head->level[i] = tail;
        }
    }

    SkipListNode* find(int val) {
        SkipListNode *p = head;
        for (int i = level - 1; i >= 0; --i) {
            while (p->level[i] && p->level[i]->val < val) {
                p = p->level[i];
            }
        }
        p = p->level[0];
        return p;
    }
    
    bool search(int target) {
        SkipListNode *p = find(target);
        return p->val == target;
    }
    
    void add(int val) {
        vector<SkipListNode *> update(MAXL);
        SkipListNode *p = head;
        for (int i = level - 1; i >= 0; --i) {
            while (p->level[i] && p->level[i]->val < val) {
                p = p->level[i];
            }
            update[i] = p;
        }
        int lv = randomLevel();
        if (lv > level) {
            update[lv] = head; 
            lv = ++level;
        }
        SkipListNode *newNode = new SkipListNode(val, lv);
        for (int i = lv - 1; i >= 0; --i) {
            p = update[i];
            newNode->level[i] = p->level[i];
            p->level[i] = newNode;
        }
        ++length;
    }
    
    bool erase(int val) {
        vector<SkipListNode *> update(MAXL + 1);
        SkipListNode *p = head;
        for (int i = level - 1; i >= 0; --i) {
            while (p->level[i] && p->level[i]->val < val) {
                p = p->level[i];
            }
            update[i] = p;
        }
        p = p->level[0];
        if (p->val != val) return false;
        for (int i = 0; i < level; ++i) {
            if (update[i]->level[i] != p) {
                break;
            }
            update[i]->level[i] = p->level[i];
        }
        while (level > 0 && head->level[level - 1] == tail) --level;
        --length;
        return true;
    }

    int randomLevel() {
        int lv = 1;
        while (lv < MAXL && (rand() & S) < PS) ++lv;
        return lv;
    }
};
